src/sampling.jl

```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
 end
 seqsample_c!(a::AbstractArray, x::AbstractArray) = seqsample_c!(Random.GLOBAL_RNG, a, x)
 
-## TODO: implement Algorithm D (page 716 - 717)
+"""
+    seqsample_d!([rng], a::AbstractArray, x::AbstractArray)
+Random subsequence sampling using algorithm D described in the following paper (page 716-17):
+Jeffrey Scott Vitter. "Faster Methods for Random Sampling". Communications of the ACM,
+27 (7), July 1984.
+This algorithm consumes ``O(k)`` random numbers, with `k=length(x)`.
+The outputs are ordered.
+"""
+function seqsample_d!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
+    N = length(a)
+    n = length(x)
+    n <= N || error("length(x) should not exceed length(a)")
+
+    i = 0
+    j = 0
+
+    vprime = exp(log(rand(rng)) / n)
+    q1 = N - n + 1
+    q2 = q1 / N
+    alpha = 1 / 13 # choose alpha value
+    threshold = alpha * n
+    X = nothing
+
+    while n > 1 && threshold < N
+
+        while true
+
+            while true
+                X = N * (1 - vprime)
+                s = trunc(Int, X)
+                if s < q1
+                    break
+                end
+                vprime = exp(log(rand(rng)) / n)
+            end
+
+            y = rand(rng) / q2
+            lhs = exp(log(y) / (n - 1))
+            rhs = ((q1 - s) / q1) * (N / (N - X))
+
+            if lhs <= rhs
+                vprime = lhs / rhs
+                break
+            end
+
+            if n - 1 > s
+                bottom = N - n
+                limit = N - s
+            else
+                bottom = N - s - 1
+                limit = q1
+            end
+
+            top = N - 1
+
+            while top >= limit
+                y = y * top / bottom
+                bottom -= 1
+                top -= 1
+            end
+
+            if exp(log(y) / (n - 1)) < N / (N - X)
+                vprime = exp(log(rand(rng)) / (n - 1))
+                break
+            end
+            vprime = exp(log(rand(rng)) / n)
+        end
+
+        @inbounds x[j+=1] = a[i+=s+1]
```

Comment:
I am not smart enough for increments in index expressions ;-)

Reply:
Does this seem fine?
```suggestion
         j += 1
         i += s+1
        @inbounds x[j] = a[i]
```